here-naksha-lib-psql/src/jvmMain/resources/naksha.sql

```diff
@@ -527,136 +151,60 @@ $$;
-  let i = 0;
-  builder.startTags();
-  while (i < keys.length) {
-    builder.writeTag()
+  if (typeof require !== "function") {
+    plv8.find_function("es_modules_init")();
+    if (typeof require !== "function") {
+      plv8.elog(ERROR, "Failed to initialize module system");
+    }
   }
-  let opCode = jb.XyzOp.Companion.getOpCode(op["op"]);
-  return builder.buildXyzOp(opCode, op["id"], op["uuid"]);
-$$;
-
-CREATE OR REPLACE FUNCTION feature_id(feature bytea) RETURNS text
-LANGUAGE 'plv8'
-IMMUTABLE
-PARALLEL SAFE
-SET search_path FROM CURRENT
-AS $$
-  let naksha = require("naksha");
-  let session = naksha.NakshaSession.Companion.get();
-  return session.getFeatureId(feature);
-$$;
-
-CREATE OR REPLACE FUNCTION feature_type(feature bytea) RETURNS text
-LANGUAGE 'plv8'
-IMMUTABLE
-PARALLEL SAFE
-SET search_path FROM CURRENT
-AS $$
-  let naksha = require("naksha");
-  let session = naksha.NakshaSession.Companion.get();
-  return session.getFeatureType(feature);
-$$;
-
-CREATE OR REPLACE FUNCTION jsonb_to_feature(feature jsonb) RETURNS bytea
-LANGUAGE 'plv8'
-IMMUTABLE
-PARALLEL SAFE
-SET search_path FROM CURRENT
-AS $$
-  return require("jbon").JbBuilder.Companion.create(1000).buildFeatureFromMap(feature);
-$$;
-
-CREATE OR REPLACE FUNCTION json_to_feature(feature text) RETURNS bytea
-LANGUAGE 'plv8'
-IMMUTABLE
-PARALLEL SAFE
-SET search_path FROM CURRENT
-AS $$
-  return require("jbon").JbBuilder.Companion.create(1000).buildFeatureFromMap(JSON.parse(feature));
-$$;
-
-CREATE OR REPLACE FUNCTION feature_to_jsonb(feature bytea) RETURNS jsonb
-LANGUAGE 'plv8'
-IMMUTABLE
-PARALLEL SAFE
-SET search_path FROM CURRENT
-AS $$
-  let jb = require("jbon");
-  let reader = new jb.JbMapFeature();
-  reader.mapBytes(feature);
-  let map = reader.root().toIMap();
-  map["id"] = reader.id();
-  return map;
+  const { Platform } = require("naksha_base");
+  const { JbFeatureDecoder } = require("naksha_jbon");
+  let decoder = new JbFeatureDecoder();
+  decoder.mapBytes(jbon);
+  return Platform.toJSON(decoder.toMap());
 $$;
 
-CREATE OR REPLACE FUNCTION feature_to_json(feature bytea) RETURNS text
-LANGUAGE 'plv8'
-IMMUTABLE
-PARALLEL SAFE
-SET search_path FROM CURRENT
-AS $$
-  let jb = require("jbon");
-  let reader = new jb.JbMapFeature();
-  reader.mapBytes(feature);
-  let map = reader.root().toIMap();
-  map["id"] = reader.id();
-  return JSON.stringify(map);
-$$;
+CREATE OR REPLACE FUNCTION naksha_jbon_map_to_jsonb(jbon bytea) RETURNS jsonb
+LANGUAGE 'plpgsql'
+IMMUTABLE PARALLEL SAFE STRICT
+AS $$ BEGIN
+  RETURN naksha_jbon_map_to_json(jbon)::jsonb;
+END $$;
 
-CREATE OR REPLACE FUNCTION row_to_ns(created_at int8, updated_at int8, txn int8, action int2, version int4, author_ts int8,
- uid int4, app_id text, author text, geo_grid int4, puid int4, ptxn int8, collection_id text) RETURNS bytea
-LANGUAGE 'plv8'
-IMMUTABLE
-PARALLEL SAFE
-SET search_path FROM CURRENT
+CREATE OR REPLACE FUNCTION naksha_tags(tags bytea, flags int4) RETURNS jsonb
```

Comment:
I wonder if there's an option to **atomically** test such functions as this one?
By atomically I mean without setting up any additional resources (like creating tables, writing collections and using other parts of our "regular" flow).
I was thinking of executing raw SQL on test DB and asserting the stdout or something.
WDYT?

Reply:
It should be possible to write tests for it, a matter of time and effort. Right now I skipped this, due to a lack of time, as priority is right now getting storage-api integration ready.